From c4bcb7564abb948d941bfcb3e3f3e40cb37fa172 Mon Sep 17 00:00:00 2001
From: Christian Marangi <ansuelsmth@gmail.com>
Date: Fri, 5 Jun 2026 17:44:11 +0200
Subject: [PATCH] PCI: mediatek: handle reset for AN7583 startup

Handle reset from SCU register for AN7583 SoC in startup OP PCIe driver.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
---
 drivers/pci/controller/pcie-mediatek.c | 32 +++++++++++++++++++++++++-
 1 file changed, 31 insertions(+), 1 deletion(-)

diff --git a/drivers/pci/controller/pcie-mediatek.c b/drivers/pci/controller/pcie-mediatek.c
index 4bebf968e65e..a0fe2d0633b8 100644
--- a/drivers/pci/controller/pcie-mediatek.c
+++ b/drivers/pci/controller/pcie-mediatek.c
@@ -30,6 +30,13 @@
 
 #include "../pci.h"
 
+/* SCU */
+#define SCU_PCIC			0x88
+#define SCU_PCIC_PERSTOUT2		BIT(16)
+
+#define SCU_RSTCTRL2			0x830
+#define SCU_RSTCTRL2_PCIE2		BIT(27)
+
 /* PCIe shared registers */
 #define PCIE_SYS_CFG		0x00
 #define PCIE_INT_ENABLE		0x0c
@@ -231,6 +238,8 @@ struct mtk_pcie {
 
 	struct list_head ports;
 	const struct mtk_pcie_soc *soc;
+
+	struct regmap *scu;
 };
 
 static void mtk_pcie_subsys_powerdown(struct mtk_pcie *pcie)
@@ -673,6 +682,16 @@ static int mtk_pcie_startup_port_v2(struct mtk_pcie_port *port)
 	u32 val;
 	int err;
 
+	if (device_is_compatible(pcie->dev, "airoha,an7583-pcie")) {
+		pcie->scu = syscon_regmap_lookup_by_phandle_optional(pcie->dev->of_node,
+								"airoha,scu");
+		if (pcie->scu) {
+			regmap_set_bits(pcie->scu, SCU_RSTCTRL2, SCU_RSTCTRL2_PCIE2);
+
+			regmap_clear_bits(pcie->scu, SCU_PCIC, SCU_PCIC_PERSTOUT2);
+		}
+	}
+
 	entry = resource_list_first_type(&host->windows, IORESOURCE_MEM);
 	if (entry)
 		mem = entry->res;
@@ -750,6 +769,12 @@ static int mtk_pcie_startup_port_v2(struct mtk_pcie_port *port)
 	val = PCIE2AHB_SIZE | WIN_ENABLE;
 	writel(val, port->base + PCIE_AXI_WINDOW0);
 
+	if (device_is_compatible(pcie->dev, "airoha,an7583-pcie") && pcie->scu) {
+		regmap_clear_bits(pcie->scu, SCU_RSTCTRL2, SCU_RSTCTRL2_PCIE2);
+
+		regmap_set_bits(pcie->scu, SCU_PCIC, SCU_PCIC_PERSTOUT2);
+	}
+
 	return 0;
 }
 
-- 
2.53.0

